Cleanup/swiftify post service

nheagy opened the pull request:
Converts PostService into swift.

To test: this potential deals with: creating, updating, trashing, restoring, uploading, and syncing posts. Try to do those things in the app—do they still work?

I tried to stick to the Swift 3 [API design guidelines](https://swift.org/documentation/api-design-guidelines/), such as naming creation methods 'make' and omitting type information from method names. Some of this is different than our existing code (we've predominantly used 'create' in the past). 

Keeping a decent Objective-C API has been a wild internal debate. Let me know what you think.

Needs review: @kurzee 
_I choo-choo-choose you!_

kurzee commented:
Added some labels and the milestone for you 🙂

kurzee commented:
Looks like a few trailing whitespace violations. 🚓🚓🚓🚓🚓🚓

kurzee commented:
Crash on launch:

```
* thread #1: tid = 0x2ffb63, 0x0000000115258805 libobjc.A.dylib`objc_msgSend + 5, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x4a4cac)
    frame #0: 0x0000000115258805 libobjc.A.dylib`objc_msgSend + 5
    frame #1: 0x0000000110616b76 Foundation`ResolvePredicateArgument + 2112
    frame #2: 0x00000001105e6e7a Foundation`_qfqp2_performParsing + 4700
    frame #3: 0x00000001105e5bb7 Foundation`+[NSPredicate predicateWithFormat:arguments:] + 46
    frame #4: 0x0000000116419d8a libswiftFoundation.dylib`(extension in Foundation):__ObjC.NSPredicate.init (format : Swift.String, Swift.Array<Swift.CVarArgType>...) -> __ObjC.NSPredicate + 154
  * frame #5: 0x000000010ddbdc55 WordPress`PostService.merge(remotePosts=14 values, type=post, statuses=2 values, authorID=4869292, blog=0x00007fa342d9a000, purge=true, completion=0x000000010ddc8610 WordPress`partial apply forwarder for WordPress.PostService.(sync (WordPress.PostService.PostType, blog : __ObjC.Blog, options : Swift.Optional<__ObjC.PostServiceSyncOptions>, success : Swift.Optional<(Swift.Array<__ObjC.AbstractPost>) -> ()>, failure : Swift.Optional<(Swift.Optional<__ObjC.NSError>) -> ()>) -> ()).(closure #1).(closure #1).(closure #1) at PostService.swift, self=0x00007fa3457826b0) -> ()) -> () + 1605 at PostService.swift:363
    frame #6: 0x000000010ddca5a3 WordPress`PostService.(self=0x00007fa3457826b0, blogID=0xd000000000700000, remotePosts=14 values, type=post, options=0x00007fa347b62f00, success=nil) -> ()?, failure : (NSError?) -> ()?) -> ()).(closure #1).(closure #1) + 1603 at PostService.swift:137
    frame #7: 0x000000010ddc855e WordPress`partial apply for PostService.(sync(PostService.PostType, blog : Blog, options : PostServiceSyncOptions?, success : ([AbstractPost]) -> ()?, failure : (NSError?) -> ()?) -> ()).(closure #1).(closure #1) + 206 at PostService.swift:0
    frame #8: 0x000000010dcf0f37 WordPress`thunk + 39 at PageListViewController.swift:0
    frame #9: 0x0000000113ed3127 CoreData`developerSubmittedBlockToNSManagedObjectContextPerform + 199
    frame #10: 0x00000001165aa3eb libdispatch.dylib`_dispatch_client_callout + 8
    frame #11: 0x00000001165921ef libdispatch.dylib`_dispatch_main_queue_callback_4CF + 1738
    frame #12: 0x00000001157280f9 CoreFoundation`__CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
    frame #13: 0x00000001156e9b99 CoreFoundation`__CFRunLoopRun + 2073
    frame #14: 0x00000001156e90f8 CoreFoundation`CFRunLoopRunSpecific + 488
    frame #15: 0x0000000118122ad2 GraphicsServices`GSEventRunModal + 161
    frame #16: 0x00000001114d8f09 UIKit`UIApplicationMain + 171
    frame #17: 0x000000010da27e9e WordPress`main(argc=1, argv=0x00007fff521de668) + 62 at main.m:5
    frame #18: 0x00000001165de92d libdyld.dylib`start + 1
    frame #19: 0x00000001165de92d libdyld.dylib`start + 1
```

kurzee commented:
Same crash mentioned above happened again when navigation from **Blog Posts** < **BlogDetails** < **BlogList**.

```
* thread #1: tid = 0x301771, 0x000000010bd63805 libobjc.A.dylib`objc_msgSend + 5, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x4a4cac)
    frame #0: 0x000000010bd63805 libobjc.A.dylib`objc_msgSend + 5
    frame #1: 0x0000000107121b76 Foundation`ResolvePredicateArgument + 2112
    frame #2: 0x00000001070f1e7a Foundation`_qfqp2_performParsing + 4700
    frame #3: 0x00000001070f0bb7 Foundation`+[NSPredicate predicateWithFormat:arguments:] + 46
    frame #4: 0x000000010cf24d8a libswiftFoundation.dylib`(extension in Foundation):__ObjC.NSPredicate.init (format : Swift.String, Swift.Array<Swift.CVarArgType>...) -> __ObjC.NSPredicate + 154
  * frame #5: 0x00000001048c8c55 WordPress`PostService.merge(remotePosts=9 values, type=post, statuses=2 values, authorID=4869292, blog=0x00007ffee2f7fe90, purge=true, completion=0x00000001048d3610 WordPress`partial apply forwarder for WordPress.PostService.(sync (WordPress.PostService.PostType, blog : __ObjC.Blog, options : Swift.Optional<__ObjC.PostServiceSyncOptions>, success : Swift.Optional<(Swift.Array<__ObjC.AbstractPost>) -> ()>, failure : Swift.Optional<(Swift.Optional<__ObjC.NSError>) -> ()>) -> ()).(closure #1).(closure #1).(closure #1) at PostService.swift, self=0x00007ffee7de7af0) -> ()) -> () + 1605 at PostService.swift:363
    frame #6: 0x00000001048d55a3 WordPress`PostService.(self=0x00007ffee7de7af0, blogID=0xd000000001780002, remotePosts=9 values, type=post, options=0x00007ffee7df9140, success=nil) -> ()?, failure : (NSError?) -> ()?) -> ()).(closure #1).(closure #1) + 1603 at PostService.swift:137
    frame #7: 0x00000001048d355e WordPress`partial apply for PostService.(sync(PostService.PostType, blog : Blog, options : PostServiceSyncOptions?, success : ([AbstractPost]) -> ()?, failure : (NSError?) -> ()?) -> ()).(closure #1).(closure #1) + 206 at PostService.swift:0
    frame #8: 0x00000001047fbf37 WordPress`thunk + 39 at PageListViewController.swift:0
    frame #9: 0x000000010a9de127 CoreData`developerSubmittedBlockToNSManagedObjectContextPerform + 199
    frame #10: 0x000000010d0b53eb libdispatch.dylib`_dispatch_client_callout + 8
    frame #11: 0x000000010d09d1ef libdispatch.dylib`_dispatch_main_queue_callback_4CF + 1738
    frame #12: 0x000000010c2330f9 CoreFoundation`__CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
    frame #13: 0x000000010c1f4b99 CoreFoundation`__CFRunLoopRun + 2073
    frame #14: 0x000000010c1f40f8 CoreFoundation`CFRunLoopRunSpecific + 488
    frame #15: 0x000000010ec2dad2 GraphicsServices`GSEventRunModal + 161
    frame #16: 0x0000000107fe3f09 UIKit`UIApplicationMain + 171
    frame #17: 0x0000000104532e9e WordPress`main(argc=1, argv=0x00007fff5b6d3668) + 62 at main.m:5
    frame #18: 0x000000010d0e992d libdyld.dylib`start + 1
    frame #19: 0x000000010d0e992d libdyld.dylib`start + 1
```

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+
+    typealias SuccessPosts = ([AbstractPost]) -> Void
+    typealias FailureBasic = (NSError?) -> Void
+    typealias SuccessBasic = () -> Void
```
I think we might avoid having a "basic" completion type in this case, and instead just default to a `() -> Void` where needed.

kurzee commented:
We lost all of our wonderful documentation with this conversion, we'll definitely need all of that added back and adapted for any swift conversion intricacies. Both from the previous `PostService.h` and `PostService.m`.

kurzee commented:
We'll need functions/method designated as `private` where they were previously private to the .m, or instead add them as private extensions where applicable if you'd like.

kurzee commented:
Just about any new public `func` should have documentation along with it, unless super obvious.

kurzee commented on WordPress/Classes/Services/PostService.h:
```diff
@@ -1,128 +1,7 @@
+typedef NSString * PostServiceType;
+extern NSString * const PostServiceTypePost;
+extern NSString * const PostServiceTypePage;
+extern NSString * const PostServiceTypeAny;
```
Not sure why we're hanging to both the .h and .m with the duplicate constants?

nheagy replied:
Because swift can't make global variables visible to Objective-C, and ObjC can't see the swift String enum. There's a solution in Xcode 8 that allows Objective-C NSString typedefs to be compiled into strongly typed swift structs, but I couldn't find anything so elegant that works _now_. So I brought back the old PostServiceType (from a few revisions ago) and added an objective-C compatible method in an extension.

I'm open to alternate ideas on this.

kurzee replied:
Ah, how about a helper method in `PostService` that exports the string value? Or instead convert `PostService.h` to something like `PostServiceTypes.h` and then import into Swift?

Just want to avoid having both two "PostServices" and maintaining duplicate string values.

kurzee commented:
If you do a copy + paste of all the code in `PostService.swift`, you'll see Xcode modify with some its own default formatting. Probably worthwhile to roll with it or at least verify.

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+    }
+
+    typealias SuccessPosts = ([AbstractPost]) -> Void
+    typealias FailureBasic = (NSError?) -> Void
```
Probably drop the word "Basic" here as it doesn't really contribute to the reusability. If we end up changing this block we would then have to rename it from "Basic".

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+        return post
+    }
+
+    func makeDraftPage(for blog: Blog) -> Page? {
```
Should probably not use `for` on `makeDraftPage` and `makeDraftPost`.

kurzee replied:
Maybe better as `func makeDraftPage(blog: Blog)`.

kurzee commented:
I see a few method calls where we've replaced `createDraftPostInMainContextForBlog` with a general call to `makeDraftPostFor()`.  (ex: https://github.com/wordpress-mobile/WordPress-iOS/pull/5899/files/2abe61ae5ad50355f99bb47db372e11f5c0a99b4#diff-0fee1392854efbd0a02ea1234101e680L136)

Before, it appears very intentional to explicitly request creating a new `Post` or `Page` on the main context for `blog`. However with these changes, this would now lead to an assertion of `assert(self.managedObjectContext == blog.managedObjectContext, "Blog's context should be the the same as the service's")` if the method caller's `blog` is not already on the main context.

I don't see any changes addressing this otherwise, so I'm thinking there are unknown (to us) scenarios in which we still do need to `createDraftPostInMainContextForBlog` and not lead to an assertion, or crash.

I may be wrong, but I think we'll need to include a swift version of `createDraftPostInMainContextForBlog:`.

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+        postPost.postType = Post.typeDefaultIdentifier
+
+        let postCategoryService = PostCategoryService(managedObjectContext:self.managedObjectContext)
+        if let category = postCategoryService.findWithBlogObjectID(blog.objectID, andCategoryID: (blog.settings?.defaultCategoryID)!) {
```
Should handle a case in which `blog.settings` or `defaultCategoryID` might be `nil`?

I'm thinking this might be a race condition between a blog sync finishing (or error) vs this method being called along the way, causing a crash.

nheagy replied:
Yeah, I agree we should.

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+        return page
+    }
+
+    func get(postID: Int, blog: Blog, success: (AbstractPost?) -> Void, failure: ((NSError?) -> Void)?) {
```
Could use `FailureBasic` or the like with `((NSError?) -> Void)?)`.

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+        return posts.first as? AbstractPost
+    }
+
+    func remoteSyncParameters(for remote: PostServiceRemote, with options: PostServiceSyncOptions?) -> [String: AnyObject]? {
```
`for`

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+        return remotePost
+    }
+
+    func remoteCategories(for post: Post) -> [RemotePostCategory]? {
```
`for`

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+        return remoteCategory
+    }
+
+    func remoteMetadata(for post: Post) -> [[String: AnyObject]] {
```
`for`

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+        }
+    }
+
+    func updateComments(for post: AbstractPost) {
```
`for`

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+        self.init(managedObjectContext: ContextManager.sharedInstance().mainContext)
+    }
+
+    func makePost(for blog: Blog) -> Post? {
```
`for`

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+        return make(type: Post.self, blog: blog)
+    }
+
+    func makePage(for blog: Blog) -> Page? {
```
`for`

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+            remote.getPostWithID(postID,
+                                 success: { (remotePost: RemotePost?) -> Void in
+                                    self.managedObjectContext.performBlock({
+                                        guard let blogInContext = (try? self.managedObjectContext.existingObjectWithID(blogID)) as? Blog else { return }
```
Let's go ahead and bring the `return` to the next line.

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,599 @@
+                                        }
+                                        ContextManager.sharedInstance().saveContext(self.managedObjectContext)
+
+                                        success(post)
```
Missing a check for `success` being `nil` here, I think?

nheagy commented:
Thanks for the review, great feedback 💯 

- Looking at the crash right now. It's going to need some more optionals/nullables :(
- Good catch on the documentation. It's on my checklist and I missed it. Spent all my time looking at the `.m`.
- The `+createDraftPostInMainContextForBlog` method(s) seemed to just be a convenience for calling the `serviceWithMainContext` constructor, which itself was just a convenience for initializing with `[[ContextManager sharedInstance] mainContext]]`. I just made that a convenience `init()` in swift, and so calling `PostService().makeDraftPost(for:blog)` or `[[PostService new] makeDraftPostFor:blog]` should be identical behavior. I can add back the class methods if you think it's important, but looking at all uses, it seems to just be to save typing?
- Formatting/whitespace: 😱 
- You don't like `for` as an argument name? I debated `for`, `forBlog`, and `blog`.

kurzee commented:
>- You don't like `for` as an argument name? I debated `for`, `forBlog`, and `blog`.

Well, it's more-so that it's technically restricted for a `for` loop, so the editor highlights it differently.

kurzee commented:
> I just made that a convenience init() in swift, and so calling PostService().makeDraftPost(for:blog) or [[PostService new] makeDraftPostFor:blog] should be identical behavior.

Hmm, I'm not sure on this one. Something feels strange here since `LocalCoreDataService` generally expects an explicit argument for which context to use. I see how it does do the same thing here, but I worry about the readability of which context `PostService` is being initialized on.

I think it might be best to mirror the original implementation a bit more for this conversion for clarity, with future refactors taking into account convenience methods that could lessen the method names.

nheagy commented:
Ya, Xcode isn't great at highlighting `for` as a parameter name, though it's a bit better in Xcode 8 it seems (only highlights it when called, not in the `func` declaration). So I'm up for changing it if it's just too ugly. FWIW it's technically ok for use as a param name, as part of [SE-0001](https://github.com/apple/swift-evolution/blob/master/proposals/0001-keywords-as-argument-labels.md) ¯\_(ツ)_/¯

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+
+    func trashRemotePost(withPost post: AbstractPost, success: SuccessBasic, failure: FailureBasic) {
+        let postObjectID = post.objectID
+        let remotePost = self.getRemotePost(withPost: post)
```
No need to use *self*. You'd only need that when referencing a method / ivar from inside a block.

(Please, review the whole PR!)

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+    func trashRemotePost(withPost post: AbstractPost, success: SuccessBasic, failure: FailureBasic) {
+        let postObjectID = post.objectID
+        let remotePost = self.getRemotePost(withPost: post)
+        if let remote = self.remote(for: post.blog) {
```
Please, avoid stacking nested levels, since it's very very hard to read the code, otherwise.

`guard let` FTW!

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+        - blog: The blog that has the post
+        - success: A success block
+        - failure: A failure block
+     */
```
Please, check out the *Swift* styled comments we're using (IE Notification.swift should be full of it). Should look like this:

```
/// Some method description
/// - parameters:
///  - somethingID: some description
///
```

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+    enum PostType: String {
+        case any = "any"
+        case post = "post"
+        case page = "page"
```
(AFAIK) Enum Cases should be uppercase. @diegoreymendez please, correct me if not, because otherwise i'd need to update few things all over!

diegoreymendez replied:
They're upper case in Swift 2 (current), and will turn lowercase with Swift 3 (future).

jleandroperez replied:
This changes everything 😄 

Thanks Diego!!!

jleandroperez replied:
By all means, apologies about the noise!!!

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+    }
+
+    func makeDraftPost(for blog: Blog) -> Post? {
+        guard let post = self.makePost(for: blog) else { return nil }
```
Please, promote the `return nil` to a new line, for readability.

(Plus, please, also nuke the `self.` reference!)

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+         - failure: A failure block
+    */
+    func sync(type: PostType, blog: Blog, options: PostServiceSyncOptions?, success: SuccessPosts?, failure: FailureBasic?) {
+        if let remote = self.remote(for:blog) {
```
Same as above, please, minimize the nested levels

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+                                    purge: options?.purgesLocalSync,
+                                    completion: { posts in
+                                        if let success = success {
+                                            success(posts)
```
You can replace the entire if statement with `success?(posts)`

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+                                })
+                            }
+                        } catch let error as NSError? {
+                            DDLogSwift.logError(String("Could not retrieve blog in context with error: %@", error))
```
Swift style:

`DDLogSwift.logError("Could not retrieve blog in context with error: \(error)")`

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+                        }
+                    })
+                }, failure: { (error) in
+                    self.managedObjectContext.performBlock({
```
When a closure is the last method's parameter, you can do this:

```
self.managedObjectContext.performBlock() {
   failure?(error)
}
```

(It can live as a "tail", no need to actually send it over as a parameter)

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+         - post: The post or page to delete
+         - success: A success block
+         - failure: A failure block
+     */
```
Same as above, please, update the comment style

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+                            ContextManager.sharedInstance().saveContext(self.managedObjectContext)
+                        }
+                    } catch let error as NSError? {
+                        DDLogSwift.logError(String("%@", error))
```
`DDLogSwift.logError("\(error)")`

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+                            postInContext.restorableStatus = nil
+                        }
+                    } catch let error as NSError? {
+                        DDLogSwift.logError(String("%@", error))
```
`DDLogSwift.logError("\(error)")`

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+        }
+
+        let failureBlock: (NSError?) -> Void = { error in
+            self.managedObjectContext.performBlock({
```
Same as above, no need to send the block inside the parenthesis:

```
moc.performBlock() {

}
```

jleandroperez commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,673 @@
+    func delete(post: AbstractPost, success: SuccessBasic, failure: FailureBasic) {
+
+        let deleteBlock: SuccessBasic = {
+            if let postID = post.postID where postID.longLongValue > 0 {
```
`guard let else` perhaps (-- nested levels --)

kurzee commented:
We're going to have to figure out how post types are to be handled. We still have two sets of the same values in `PostServiceTypes.h/m` and the `PostService` enum.

Particularly, as mentioned before, the post type used in `PostService` is more of an API string value that can be could also be a custom string. For example, we provide an enum value of `PostTypeAny` that doesn't really make sense as an object type. Which is why these were originally named as `PostServiceType` versus `PostType` that reads as an object type. This also collides with the model object `PostType` itself, which actually encapsulates the usage of a post type.

Menus passes in custom post type strings to query posts of a custom content type. We'll also be extending our Editor in the near future to support custom types as now supported on the web.

Also, Menus is still crashing as it was not adopted to use the new swift `PostService`. Please search and scan the rest of the project for any other implementations of the original `PostService` that might also be broken.

I think ultimately we need two things:

- A single place for the `post`, `page`, and `any` string values.
- Allow `PostService` to accept a `String`, and let the users provide the source of that string value.

kurzee commented:
>Ya, Xcode isn't great at highlighting for as a parameter name

I'd say let's avoid using `for` until Xcode can really handle it. Makes it easier for reading and understanding the code.

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,678 @@
+    ///  - parameter success: A success block
+    ///  - parameter failure: A failure block
+    func sync(type: PostType, blog: Blog, options: PostServiceSyncOptions?, success: SuccessPostsBlock?, failure: FailureBlock?) {
+        if let remote = remote(for:blog) {
```
Can we use a `guard` here instead, to avoid wrapping this whole thing in `{}`.

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,678 @@
+    ///  - parameter failure: A failure block
+    func sync(type: PostType, blog: Blog, options: PostServiceSyncOptions?, success: SuccessPostsBlock?, failure: FailureBlock?) {
+        if let remote = remote(for:blog) {
+            let blogID = blog.objectID
```
Let's go with the original `blogObjectID` to differentiate.

kurzee commented on WordPress/Classes/Services/PostService.swift:
```diff
@@ -0,0 +1,678 @@
+            remote.getPostsOfType(type.rawValue, options: remoteOptions, success: { (remotePosts) in
+                    self.managedObjectContext.performBlock() {
+                        do {
+                            if let blogInContext = try self.managedObjectContext.existingObjectWithID(blogID) as? Blog {
```
Hmm, is the `if let` necessary here? I _think_ the error would be caught and break the execution?

kurzee replied:
This also may not handle the similar case in which `blogInContext` was `nil` but should still log the error.

nheagy commented on WordPress/Classes/Networking/Remote Objects/RemotePost.h:
```diff
@@ -34,8 +34,9 @@
+ // typed Dict to [String : String]. Not sure if that should be [String : AnyObject] -nh
  */
-@property (nonatomic, strong) NSArray *metadata;
+@property (nonatomic, strong) NSArray<NSDictionary<NSString *, id> *> *metadata;
```
`NSDictionary` can use any object as a key?

kurzee replied:
It can (NSCopying), but just not sure we gain much with this particular use of generics.

nheagy commented:
@kurzee here's what's been done:
- fixed formatting and whitespace issues
- fixed crashes you had mentioned
- added inline documentation back, then updated its formatting to match Xcode 8 style
- added back the '…InMainContext…' creation class methods
- marked methods private if they had been hidden in the Objective-C
- renamed PostService.h/m to PostServiceTypes.h/m
- removed function calls that use self
- removed trailing closures in parens
- removed non-swifty log calls
- added handling of nil blog.settings or defaultCategoryID
- moved returns onto next line
- changed `if let success = success…` to `success?()` 

The rest:
### For
I changed the `for blog:` parameters to just `blog:` in the make* methods because of the swift 3 guideline "Initializer and factory method calls should form a phrase that does not include the first argument". I left the others because I still feel this is the intended result of the API guidelines. If the syntax highlighting in Xcode is a breaking point, do you think we should do `blog:` or `forBlog blog:`?

### I'm not able to get menus to crash.
I played with this a few times this week, but I'm not very familiar with this functionality. Can you provide steps to repeat it?

### Stringly-typed PostTypes
We can resolve the duplication of this type in Xcode 8, but there is more to this?

Changing the name of `PostServiceType` was driven by [this comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/5655#discussion_r71910571) on a previous PR. I'm not familiar with the menus-related changes, but you're saying that there are types with string values beyond the existing `post`, `page`, and `any`?

Granted I've only briefly considered this, but it seems that would warrant a deeper refactoring of the way we describe post types? I'm thinking of a class to represent types that has methods for comparison, etc. In swift I'd imagine an enum case with an associated value so that we can still have strongly typed parameters, such as:

```
enum PostTypes {
    case .any
    case .post
    case .page
    case .custom(String)
    /// …
}
```

Thoughts?

kurzee commented:
>`case .custom(String)`

I think that would probably do the trick right there.

kurzee commented:
>I played with this a few times this week, but I'm not very familiar with this functionality. Can you provide steps to repeat it?

Menus is the only case, right now, that support custom post type strings. This will soon be adopted for the editor as well, now that the web/API support custom types.

In Menus, you can see how this is handled in: https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/ViewRelated/Menus/MenuItemAbstractPostsViewController.m#L59

Test by:

1. Open Menus on a site with admin access.
2. Select a menu item from the list of items for the selected menu.
3. On the modal, select the back button "<" to change the item type.
4. Listed here is the available item types.
5. To see a custom type, enable a theme on the site which uses one of the https://en.support.wordpress.com/custom-post-types/
6. Create a post that uses the custom post type.
7. Reopen menus and notice the newly available custom type following steps 1-4.
6. Select the custom type, Crash.

So ultimately, a user of `PostService` needs to be able to pass a custom string value for the type of posts to sync, create, etc, while easily supporting the default options of `post` and `page`.

nheagy commented:
This will return at a later date, in a different form :D
